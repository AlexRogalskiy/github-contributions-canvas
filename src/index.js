import moment from "moment";

function getDateInfo(data, date) {
  return data.contributions.find(contrib => contrib.date === date);
}

function getContributionCount(graphEntries) {
  return graphEntries.reduce((rowTotal, row) => {
    return (
      rowTotal +
      row.reduce((colTotal, col) => {
        return colTotal + (col.info ? col.info.count : 0);
      }, 0)
    );
  }, 0);
}

const DATE_FORMAT = "YYYY-MM-DD";
const boxWidth = 10;
const boxMargin = 2;
const textHeight = 15;
const fontFace = "IBM Plex Mono";
const headerHeight = 60;
const canvasMargin = 20;
const yearHeight = textHeight + (boxWidth + boxMargin) * 7 + canvasMargin;
const scaleFactor = window.devicePixelRatio || 1;

function drawYear(ctx, year, offsetX = 0, offsetY = 0, data) {
  const thisYear = moment().format("YYYY");
  const today = year.year === thisYear ? moment() : moment(year.range.end);
  const start = moment(`${year.year}-01-01`);
  const firstDate = start.clone();

  if (firstDate.day() !== 6) {
    firstDate.day(-(firstDate.day() + 1 % 7));
  }

  const nextDate = firstDate.clone();
  const firstRowDates = [];
  const graphEntries = [];

  while (nextDate <= today && nextDate.day(7) <= today) {
    const date = nextDate.format(DATE_FORMAT);
    firstRowDates.push({
      date,
      info: getDateInfo(data, date)
    });
  }

  graphEntries.push(firstRowDates);

  for (let i = 1; i < 7; i += 1) {
    graphEntries.push(
      firstRowDates.map(dateObj => {
        const date = moment(dateObj.date)
          .day(i)
          .format(DATE_FORMAT);
        return {
          date,
          info: getDateInfo(data, date)
        };
      })
    );
  }

  const count = new Intl.NumberFormat().format(
    getContributionCount(graphEntries)
  );

  ctx.textBaseline = "hanging";
  ctx.fillStyle = "#000000";
  ctx.font = `10px '${fontFace}'`;
  ctx.fillText(
    `${year.year}: ${count} Contribution${year.total === 1 ? "" : "s"}${
      thisYear === year.year ? " (so far)" : ""
    }`,
    offsetX,
    offsetY
  );

  for (let y = 0; y < graphEntries.length; y += 1) {
    for (let x = 0; x < graphEntries[y].length; x += 1) {
      const day = graphEntries[y][x];
      if (moment(day.date) > today || !day.info) {
        continue;
      }
      ctx.fillStyle = day.info.color;
      ctx.fillRect(
        offsetX + (boxWidth + boxMargin) * x,
        offsetY + textHeight + (boxWidth + boxMargin) * y,
        10,
        10
      );
    }
  }
}

function drawMetaData(ctx, opts = {}) {
  const { username, width, height, footerText } = opts;
  ctx.fillStyle = "#ffffff";
  ctx.fillRect(0, 0, width, height);

  if (footerText) {
    ctx.fillStyle = "#666666";
    ctx.textBaseline = "bottom";
    ctx.font = `10px '${fontFace}'`;
    ctx.fillText(footerText, canvasMargin, height - 5);
  }

  ctx.fillStyle = "#000000";
  ctx.textBaseline = "hanging";
  ctx.font = `20px '${fontFace}'`;
  ctx.fillText(`@${username} on Github`, canvasMargin, canvasMargin);

  ctx.beginPath();
  ctx.moveTo(canvasMargin, 55);
  ctx.lineTo(width - canvasMargin, 55);
  ctx.strokeStyle = "#EBEDF0";
  ctx.stroke();
}

export function drawContributions(canvas, opts) {
  const { data, username } = opts;
  const height =
    data.years.length * yearHeight + canvasMargin + headerHeight + 10;
  const width = 53 * (boxWidth + boxMargin) + canvasMargin * 2;

  canvas.width = width * scaleFactor;
  canvas.height = height * scaleFactor;

  const ctx = canvas.getContext("2d");
  ctx.scale(scaleFactor, scaleFactor);
  ctx.textBaseline = "hanging";

  drawMetaData(ctx, {
    ...opts,
    width,
    height
  });

  data.years.forEach((year, i) => {
    const offsetY = yearHeight * i + canvasMargin + headerHeight;
    const offsetX = canvasMargin;
    drawYear(ctx, year, offsetX, offsetY, data);
  });
}
