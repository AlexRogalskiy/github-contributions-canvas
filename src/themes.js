export const themes = {
  standard: {
    background: "#ffffff",
    text: "#000000",
    meta: "#666666",
    grade4: "#196127",
    grade3: "#239a3b",
    grade2: "#7bc96f",
    grade1: "#c6e48b",
    grade0: "#ebedf0"
  },
  halloween: {
    background: "#ffffff",
    text: "#000000",
    meta: "#666666",
    grade4: "#03001C",
    grade3: "#FE9600",
    grade2: "#FFC501",
    grade1: "#FFEE4A",
    grade0: "#ebedf0"
  },
  teal: {
    background: "#ffffff",
    text: "#000000",
    meta: "#666666",
    grade4: "#458B74",
    grade3: "#66CDAA",
    grade2: "#76EEC6",
    grade1: "#7FFFD4",
    grade0: "#ebedf0"
  },
  leftPad: {
    background: "#000000",
    text: "#ffffff",
    meta: "#999999",
    grade4: "#F6F6F6",
    grade3: "#DDDDDD",
    grade2: "#A5A5A5",
    grade1: "#646464",
    grade0: "#2F2F2F"
  }
};
