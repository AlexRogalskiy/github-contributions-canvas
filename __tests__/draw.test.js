import mockData from "./mock.json";
import { drawContributions } from "../src";
import { createCanvas } from "canvas";
const canvas = createCanvas(200, 200);

test("draws contributions correctly", () => {
  drawContributions(canvas, {
    data: mockData,
    username: "sallar",
    footerText: "test",
    theme: "standard"
  });
  expect(canvas.toDataURL()).toMatchSnapshot();
});
