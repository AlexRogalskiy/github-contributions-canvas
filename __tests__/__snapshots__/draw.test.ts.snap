// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draws contributions correctly 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqQAAARyCAYAAABx8eehAAAABmJLR0QA/wD/AP+gvaeTAAAenklEQVR4nO3dMWtU34PH4XGx2N1GEKvYWuQVCIKvJmCfwjJlSgt7Ia9GEHwFKWxNJYLN7nazzRbLr5B7879nPudmnqf+MtxMZpIPpznPjsfj8QAAAJF/qx8AAIDzJkgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgNTz+gEAAEb49fvPot2rly8Oh8Ph8O3np0X7d68/Hg6Hw+Hm+/Wi/e3bz4fD4XB48+X9ov2PD18Ph8PhcHF1uWj/cHe/aDczJ6QAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApJ4dj8dj/RAAAJwvJ6QAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAKnnW7/gf/33/2z9kgAATOQ//+PfN309J6QAAKSeHY/HY/0QAACcLyekAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAKnnW7/gt5+fFu3evf54kv2v338W7V+9fHGS57n5fr1of/v280meZ7b3c+37s/Z5Rr/+2p/3zZf3i/Y/PnxdtAOAPXJCCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEDq2fF4PNYPAQDA+XJCCgBA6snfZT/6bvHRd7Wv3V9cXS7aP9zdP+r1Z/t9ndvzA8BT5IQUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJS77AEASDkhBQAgtfld9qPvgh99V/ja5x99N/3o51n7+r9+/1m0f/XyxaP2s30eRj/P2vcfAJ4iJ6QAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApNxlDwBAygkpAACp/C772e4WH/08e/953WX/99df+/NeXF0u2j/c3S/aAcAeOSEFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIOUuewAAUk5IAQBIbX6X/ei7yEfv33x5v2j/48PXR73+2rvU1+5HP/9sd82P/nnX3k2/dr/2+QHgKXJCCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEDKXfYAAKSckAIAkNr8LvvRd7Wv3a+9W3zt64++q3225z+3/ej3HwBwQgoAQEyQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBAyl32AACknJACAJDa/C772e4WH/08a+8uH33X/N7vsh99d/xszwMAOCEFACAmSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIOUuewAAUk5IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIPV86xf89vPTot271x/Pcn/z/XrR/vbt58PhcDj8+v1n0f7VyxePev3Zft7ZnufNl/eL9j8+fD3J85zb53nv7+fef961zz96P9v7P/p5Zvs8r92v/fs5+u+Jz8Pfn6fmhBQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACA1LPj8XisHwIAgPPlhBQAgNTmd9mPvmt173f7zna3+Oj96LuMZ7ubfrbf795ff7bf72zvz+i7xUfftb337+Pou8tn+zzP9jyz/X5n+zyv/bzVnJACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkHKXPQAAKSekAACkdneX/Wz70Xc9z/bzzvb8a+8Onu0u8tHPP9t+9Pt5bvvZ7rLf+9+T2fZ7vzt+9N/Di6vLRfuHu/tHPc/an3ft88z2+6o5IQUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAg5S57AABSTkgBAEhtfpf9bHezntt+truwZ7u7fLb3f/T7OdvnZ7bf12y/39Gf/9F32c/2+137/KP3s70/s33fZ3ue0f8f135+Rn8eak5IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEi5yx4AgJQTUgAAUvld9vbb7tfepXtxdblo/3B3/6j9bHdzn9vd4nu/e3r072v066/9vsz292S2u9pHf/5nu1t8trvsZ3t/Rn8fz+3zXHNCCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEDKXfYAAKSckAIAkHKX/Yn3a++Wne0u6bV3f8/2/s92t/Jsdx+P/v2Ofn9G/7xr97N9H9fuZ/v7M9td6rM9z97vst/793H0+zn673PNCSkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKXfZAwCQckIKAEDKXfb/2O/9rvlz2699P2e7y3j03cTn9v7vfT/b3eijn+fcXn/0/4vR+9k+n6Off7afd/S+5oQUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJS77AEASDkhBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgNTzrV/w1+8/i3avXr44HA6Hw83360X727efT/L6F1eXi/YPd/cnef21P++3n58W7d+9/niS/ejnX/t+rt2P/nntt93P9n188+X9ov2PD18Ph8N838fRf59ne57Z3v/Z3p+1rz/b/9+138fRzz/b/+u178/WnJACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkHp2PB6P9UMAAHC+nJACAJDa/C77vd8FPPruaXcx/30/+v2c7e7g2e6C3/t+ts/DbM+z99/XbM8/2/s/+nn2fpf93r+/a/dr38+1r781J6QAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApNxlDwBAygkpAACpze+yn+2u4dnuUp/t+c/tLubZfl77bfd7//2O/v6++fJ+0f7Hh6+Pev3Rd6Of2+93trvmR3/eRu9n+z6O7o3Rv9+tOSEFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIOUuewAAUk5IAQBI5XfZj75rdfT+4upy0f7h7v5Rr7/3u55nu5t+tvfHftv9bN+v0c9zbt+Xvf892fvzr/1/N/ru9dn2b768X7T/8eHro15/9N+TtfutOSEFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIOUuewAAUk5IAQBI5XfZj76rd7a7ZWd7f2a7K3m255nt93tu+9F3u8/2fZ/tbuvZPg97/z7O9vdwtrvaRz/P6L//o59ntp93a05IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEi5yx4AgJQTUgAAUru7y370/uLqctH+4e7+cDjMd7fybM8z213na+8Cnu3nHf15OLfPz2yfz9HPM9td2Of2922293O251/7+Rz993+278vo/dr3c2tOSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBIucseAICUE1IAAFK7u8t+9F3zo+9+3ftd2Od21/Zsd0nvfT/6+zXbXeprP8+z3S0++u/PbM8/2/d99Psz+v/p2vff521fz7M1J6QAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApNxlDwBAygkpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApJ5v/YK/fv9ZtHv18oW9vf2Z7b/9/LRo/+71x5M8j/3T3r/58n7R/seHr496/Zvv14v2t28/P+r1Z/u+zPY8o3+/a1//4upy0f7h7v5RzzP687Z2vzUnpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACknh2Px2P9EAAAnC8npAAApNxlb2+/4X62u55H79ferTzb+7P2eWZ7fvtt9+f2+137846+S3229//c/j6sff6tOSEFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIOUuewAAUk5IAQBIucve3t5+2v253Z1tv+3+zZf3i/Y/Pnx91H62n3e279fe93u/y37082zNCSkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKXfZAwCQckIKAEDKXfb29n/Zu+v57/u1dyWf213h9vb28+xH300/2132o/9/bc0JKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApd9kDAJByQgoAQMpd9v/Yu2vb/pz3F1eXi/YPd/ePev3Z7nqebe/92XY/+i7vc9v7fO5rP/rv+dackAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQcpc9AAApJ6QAAKTcZf+PvbuP7c95P/qu6tGv7/tr///37l7f197fn7+//ujnqTkhBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACDlLnsAAFJOSAEASO3uLvu933W7973352nv1/5+196V7Pv79/1sd2HvfX9xdblo/3B3f5Ln2ft+9Odztr8Ps/19G72vOSEFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIOUuewAAUk5IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIPV86xf89fvPot2rly8etb+4uly0f7i7P8nz2Nvb22+19/fN3t5+L/utOSEFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIPXseDwe64cAAOB8OSEFACC1u7vs7e3PeX/z/XrR/vbt55M8z7ntvf/29vb27rIHAOAJEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJBylz0AACknpAAApNxlb29vb29vb29vv2q/NSekAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKTcZQ8AQMoJKQAAKXfZ29tvuL/5fr1of/v280mex97e3t7efsR+a05IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEi5yx4AgJQTUgAAUu6yt7e3t7e3t7c/833NCSkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKXfZAwCQckIKAEDKXfb2U+1vvl8v2t++/XyS57G3t7e3tz+Hfc0JKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApd9kDAJByQgoAQMpd9vb29vb29vb2k+9vvl8v2t++/fyo1685IQUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAg5S57AABSTkgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAg9XzrF/z1+8+i3auXL+zPYP/t56dF+3evP57keezt7cftZ/u+z/Y89n/f33y/XrS/ffv5JM9j//f91pyQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJB6djwej/VDAABwvpyQAgCQevJ32a+9y9jdx/b29vZPY+/v+b72fl/72m/NCSkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKXfZAwCQckIKAEDKXfb/x1329vb29vb29vbL9ltzQgoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBAyl32AACknJACAJB68nfZ29vb29vb29vb/31fc0IKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQMpd9gAApJyQAgCQcpe9/VT7bz8/Ldq/e/3xJM9jb28/bn9xdblo/3B3f5Lnsbc/533NCSkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKXfZAwCQckIKAEDqyd9l/+bL+0X7Hx++nuR57O3t7bfa33y/XrS/ffv5JM8zev/t56dF+3evP57keS6uLhftH+7uT/I89vb/yr7mhBQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlLvsAQBIOSEFACD15O+yt7e3t7e3H7H/9vPTov271x9P8jz29v/KvuaEFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUu+wBAEg5IQUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAIDU861f8NfvP4t2r16+sLe3t7e3t7e33+F+a05IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEg9Ox6Px/ohAAA4X05IAQBIucve3t7e3t7e3t5+1X5rTkgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASLnLHgCAlBNSAABS7rK3t7e3t7e3tz/zfc0JKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApd9kDAJByQgoAQMpd9vb29vb29k9gf3F1uWj/cHd/kuex39e+5oQUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJS77AEASDkhBQAg5S57+6n27mK2t7e3t1+yv/l+vWh/+/bzSZ5n7/uaE1IAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUu6yBwAg5YQUAIDU5nfZu4vc3t7e3v4c9rPdpe7/r/2/sq85IQUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAg5S57AABSTkgBAEhtfpf9bHez2tvb29vb29vb/31fc0IKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQMpd9gAApJyQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQOr51i/46/efRbtXL1/Y29vb29vb29vvcL81J6QAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApJ4dj8dj/RAAAJwvJ6QAAKTcZW9vb29vb29vb79qvzUnpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACk3GUPAEDKCSkAACl32dvb29vb29vb26/ab80JKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApd9kDAJByQgoAQMpd9vb2O9pfXF0u2j/c3Z/keeyf9v7bz0+L9u9efzzJ89jb28+z35oTUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABS7rIHACDlhBQAgJS77O3td7R/8+X9ov2PD19P8jz29vb29ue535oTUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABS7rIHACDlhBQAgJS77O3td7T/9vPTov271x9P8jz29k9pf3F1uWj/cHd/kuex39d+7d/nm+/Xi/a3bz8/6nncZQ8AACsIUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASLnLHgCAlBNSAABS7rK3t9/R3l329vbj9r5f9vbL91tzQgoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBAyl32AACknJACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAqf8FnsSsEv4ZIpcAAAAASUVORK5CYII="`;
